crates/miden-objects/src/block/nullifier_tree.rs

```diff
@@ -2,37 +2,74 @@ use alloc::boxed::Box;
 use alloc::string::ToString;
 use alloc::vec::Vec;
 
-use miden_core::EMPTY_WORD;
 use miden_core::utils::{ByteReader, ByteWriter, Deserializable, Serializable};
+use miden_core::{EMPTY_WORD, Felt, FieldElement, Word};
 #[cfg(feature = "std")]
 use miden_crypto::merkle::{LargeSmt, LargeSmtError, SmtStorage};
 use miden_crypto::merkle::{MerkleError, MutationSet, Smt, SmtProof};
 use miden_processor::{DeserializationError, SMT_DEPTH};
 
-use crate::Word;
 use crate::block::{BlockNumber, NullifierWitness};
 use crate::errors::NullifierTreeError;
 use crate::note::Nullifier;
 
-// FREE HELPER FUNCTIONS AND CONSTANTS
+// CONSTANTS
 // ================================================================================================
 
 /// The value of an unspent nullifier in the tree.
 pub(super) const UNSPENT_NULLIFIER: Word = EMPTY_WORD;
 
-/// Returns the nullifier's leaf value in the SMT by its block number.
-pub(super) fn block_num_to_nullifier_leaf_value(block: BlockNumber) -> Word {
-    Word::from([block.as_u32(), 0, 0, 0])
+/// A nullifier block value in the [`NullifierTree`].
+///
+/// # Invariants
+///
+/// The [`NullifierBlock`] guarantees the following:
+/// - The felt at index 0 is a valid [`BlockNumber`].
+/// - The remaining felts are set to zero.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub struct NullifierBlock(BlockNumber);
+
+impl NullifierBlock {
+    pub const UNSPENT: NullifierBlock = NullifierBlock(BlockNumber::GENESIS);
+
+    pub fn new(value: Word) -> Result<Self, NullifierTreeError> {
+        if TryInto::<u32>::try_into(value[0].as_int()).is_ok()
+            && value[1..4].iter().all(|l| l.inner() == 0)
+        {
+            return Ok(Self(BlockNumber::from(value[0].as_int() as u32)));
+        }
+        Err(NullifierTreeError::InvalidNullifierBlockNumber)
+    }
+
+    pub fn is_spent(&self) -> bool {
+        *self != Self::UNSPENT
+    }
+
+    pub fn is_consumed(&self) -> bool {
+        self.0 != BlockNumber::GENESIS
+    }
 }
 
-/// Given the leaf value of the nullifier SMT, returns the nullifier's block number.
+impl From<BlockNumber> for NullifierBlock {
+    fn from(block_num: BlockNumber) -> Self {
+        Self(block_num)
+    }
+}
+
+/// Returns the [`BlockNumber`] at which the nullifier was consumed.
 ///
-/// There are no nullifiers in the genesis block. The value zero is instead used to signal
-/// absence of a value.
-pub(super) fn nullifier_leaf_value_to_block_num(value: Word) -> BlockNumber {
-    let block_num: u32 = value[0].as_int().try_into().expect("invalid block number found in store");
+/// There are no nullifiers in the genesis block, and the block number zero is used to signal
+/// an unconsumed nullifier.
```

Comment:
We can remove the docs here and instead rewrite the docs of `NullifierBlock` to:

```rust
/// The [`BlockNumber`] at which a [`Nullifier`] was consumed.
///
/// Since there are no nullifiers in the genesis block the [`BlockNumber::GENESIS`] is used to
/// signal an unconsumed nullifier.
///
/// This type can be converted to a [`Word`] which is laid out like this:
///
/// ```text
/// [block_num, 0, 0, 0]
/// ```
```

In particular, we can remove the `Invariants` section, since the type guarantees it contains a valid `BlockNumber`.